File: go-vllm-api/internal/handler/health/health.go
package health

import (
	"fmt"
	"net/http"
	"time"

	"go-vllm-api/internal/repository"
	"go-vllm-api/internal/service"

	prom "go-vllm-api/internal/pkg/prometheus"

	"github.com/gin-gonic/gin"
)

type VLLMProxyAccessor interface {
	CircuitBreakerStats() map[string]interface{}
}

type SysCtlAccessor interface {
	IsServiceRunning(name string) bool
	GetServiceInfo(name string) map[string]string
}

type LlamaCppAccessor interface {
	GetServerStatus() map[string]service.LlamaCppProcessInfo
	IsServerRunning(modelName string) bool
}

type HealthHandler struct {
	gpuMonitor  *service.GPUMonitor
	scheduler   *service.Scheduler
	metrics     *service.MetricsCollector
	cache       *service.CacheService
	prometheus  *prom.PrometheusExporter
	vllmProxy   VLLMProxyAccessor
	sysCtl      SysCtlAccessor
	llamaCppMgr LlamaCppAccessor
	redisRepo   *repository.RedisRepo
}

func NewHealthHandler(gpuMonitor *service.GPUMonitor, scheduler *service.Scheduler, metrics *service.MetricsCollector, cache *service.CacheService, prometheus *prom.PrometheusExporter, vllmProxy VLLMProxyAccessor, sysCtl SysCtlAccessor, llamaCppMgr LlamaCppAccessor, redisRepo *repository.RedisRepo) *HealthHandler {
	return &HealthHandler{
		gpuMonitor:  gpuMonitor,
		scheduler:   scheduler,
		metrics:     metrics,
		cache:       cache,
		prometheus:  prometheus,
		vllmProxy:   vllmProxy,
		sysCtl:      sysCtl,
		llamaCppMgr: llamaCppMgr,
		redisRepo:   redisRepo,
	}
}

func (h *HealthHandler) RegisterRoutes(rg *gin.RouterGroup) {
	rg.GET("/health", h.HealthCheck)
	rg.GET("/health/detailed", h.HealthCheckDetailed)
	rg.GET("/health/history", h.GetHealthHistory)
	rg.GET("/metrics", h.GetPrometheusMetrics)
	rg.GET("/metrics/metadata", h.GetMetricsMetadata)
}

func prometheusGPUStatus(status *service.GPUStatus) *prom.GPUMetricsData {
	if status == nil {
		return nil
	}
	data := &prom.GPUMetricsData{
		GPUCount: status.GPUCount,
		GPUs:     make([]prom.GPUGPUMetricEntry, 0, len(status.AllGPUs)),
	}
	for _, gpu := range status.AllGPUs {
		data.GPUs = append(data.GPUs, prom.GPUGPUMetricEntry{
			ID:              fmt.Sprintf("%d", gpu.Index),
			TotalMemory:     gpu.TotalMemory,
			UsedMemory:      gpu.UsedMemory,
			AvailableMemory: gpu.AvailableMemory,
			Temperature:     gpu.Temperature,
			Utilization:     gpu.Utilization,
			PowerDraw:       gpu.PowerDraw,
			FanSpeed:        gpu.FanSpeed,
			ClockSM:         gpu.ClockSM,
			EccErrors:       gpu.EccErrors,
			PCIeRx:          gpu.PCIeRxThroughput,
			PCIeTx:          gpu.PCIeTxThroughput,
			Bar1Total:       gpu.Bar1TotalMemory,
			Bar1Used:        gpu.Bar1UsedMemory,
			Throttled:       len(gpu.ThrottleReasons) > 0,
		})
	}
	return data
}

func prometheusVLLMMetrics(metrics *service.VLLMMetricsData) *prom.VLLMMetricsData {
	if metrics == nil {
		return nil
	}
	return &prom.VLLMMetricsData{
		RunningRequests:      float64(metrics.RunningRequests),
		WaitingRequests:      float64(metrics.WaitingRequests),
		GPUCacheUsage:        metrics.GPUCacheUsage,
		CPUCacheUsage:        metrics.CPUCacheUsage,
		GenerationThroughput: metrics.GenerationThroughput,
		PromptThroughput:     metrics.PromptThroughput,
		TTFT:                 metrics.TimeToFirstToken,
		TPOT:                 metrics.TimePerOutputToken,
		PrefixCacheHitRate:   metrics.PrefixCacheHitRate,
	}
}

func (h *HealthHandler) buildEngineStatus() map[string]interface{} {
	engines := map[string]interface{}{}
	vllmRunning := h.sysCtl.IsServiceRunning("vllm-aiclient")
	vllmInfo := h.sysCtl.GetServiceInfo("vllm-aiclient")
	engines["vllm"] = map[string]interface{}{
		"running": vllmRunning,
		"status":  vllmInfo["status"],
		"port":    vllmInfo["port"],
	}
	llamaStatus := h.llamaCppMgr.GetServerStatus()
	llamaMap := map[string]interface{}{}
	for k, v := range llamaStatus {
		llamaMap[k] = v
	}
	engines["llama_cpp"] = llamaMap
	return engines
}

func (h *HealthHandler) buildModelStatus() map[string]interface{} {
	models := h.scheduler.GetAvailableModels()
	modelStatus := map[string]interface{}{}
	for _, m := range models {
		modelStatus[m] = map[string]interface{}{
			"running":         h.scheduler.IsModelRunning(m),
			"port":            h.scheduler.GetModelPort(m),
			"service":         h.scheduler.GetModelService(m),
			"active_requests": h.scheduler.GetActiveRequests(m),
			"preloaded":       h.scheduler.IsModelPreloaded(m),
			"backend_type":    h.scheduler.GetModelBackendType(m),
		}
	}
	return modelStatus
}

func (h *HealthHandler) buildQueueStatus() map[string]interface{} {
	models := h.scheduler.GetAvailableModels()
	queue := map[string]interface{}{}
	for _, m := range models {
		queue[m] = map[string]interface{}{
			"active_requests":   h.scheduler.GetActiveRequests(m),
			"concurrency_limit": h.scheduler.GetConcurrencyLimit(),
			"can_accept":        h.scheduler.CanAcceptRequest(m),
		}
	}
	return queue
}

func (h *HealthHandler) HealthCheck(c *gin.Context) {
	refresh := c.Query("refresh") == "true"
	cacheKey := "api:health"

	if !refresh {
		if cached := h.cache.Get(cacheKey); cached != nil {
			c.JSON(http.StatusOK, cached)
			return
		}
	}

	gpuStatus := h.gpuMonitor.GetStatus()
	vllmMetrics := h.gpuMonitor.GetVLLMMetrics()
	healthInfo := h.metrics.GetComprehensiveHealthScore(gpuStatus, vllmMetrics)
	currentModel := h.scheduler.GetCurrentModelName()

	result := gin.H{
		"status":          healthInfo["status"],
		"timestamp":       time.Now().Format(time.RFC3339),
		"health_score":    healthInfo["overall"],
		"current_model":   currentModel,
		"default_model":   h.scheduler.GetDefaultModel(),
		"switching":       h.scheduler.IsSwitchingInProgress(),
		"details":         healthInfo,
		"engines":         h.buildEngineStatus(),
		"models":          h.buildModelStatus(),
		"circuit_breaker": h.vllmProxy.CircuitBreakerStats(),
		"redis":           h.redisRepo.IsConnected(),
	}

	h.cache.Set(cacheKey, result, 5)
	c.JSON(http.StatusOK, result)
}

func (h *HealthHandler) HealthCheckDetailed(c *gin.Context) {
	cacheKey := "api:health:detailed"
	if cached := h.cache.Get(cacheKey); cached != nil {
		c.JSON(http.StatusOK, cached)
		return
	}

	gpuStatus := h.gpuMonitor.GetStatus()
	vllmMetrics := h.gpuMonitor.GetVLLMMetrics()
	healthInfo := h.metrics.GetComprehensiveHealthScore(gpuStatus, vllmMetrics)

	gpuAlerts := h.metrics.GetGPUAlerts(gpuStatus)
	alertReasons := make([]string, 0)
	for _, a := range gpuAlerts {
		alertReasons = append(alertReasons, a.Message)
	}

	var overallScore float64
	if v, ok := healthInfo["overall"]; ok {
		overallScore, _ = v.(float64)
	}
	statusStr := "healthy"
	if v, ok := healthInfo["status"]; ok {
		statusStr, _ = v.(string)
	}

	gpuAvail := gpuStatus != nil && gpuStatus.Status == "available"
	gpuUtilization := 0
	gpuTemp := 0
	gpuMemUsedPct := 0
	if gpuStatus != nil {
		gpuUtilization = gpuStatus.Utilization
		gpuTemp = gpuStatus.Temperature
		gpuMemUsedPct = gpuStatus.MemoryUtilization
	}

	vllmRunning := h.sysCtl.IsServiceRunning("vllm-aiclient")
	vllmActiveRequests := 0
	if vllmMetrics != nil {
		vllmActiveRequests = vllmMetrics.RunningRequests
	}

	redisAvailable := h.redisRepo.IsConnected()
	redisConnected := h.redisRepo.IsConnected()

	result := gin.H{
		"overall_score": overallScore,
		"status":        statusStr,
		"checks": gin.H{
			"gpu": gin.H{
				"available":       gpuAvail,
				"utilization":     gpuUtilization,
				"temperature":     gpuTemp,
				"memory_used_pct": gpuMemUsedPct,
			},
			"go_backend": gin.H{
				"reachable":        true,
				"response_time_ms": 0,
			},
			"python_backend": gin.H{
				"reachable":        true,
				"response_time_ms": 0,
			},
			"vllm_service": gin.H{
				"running":         vllmRunning,
				"active_requests": vllmActiveRequests,
			},
			"redis": gin.H{
				"available": redisAvailable,
				"connected": redisConnected,
			},
		},
		"alert_reasons": alertReasons,
		"timestamp":     time.Now().Format(time.RFC3339),
	}

	h.cache.Set(cacheKey, result, 5)
	c.JSON(http.StatusOK, result)
}

func (h *HealthHandler) GetPrometheusMetrics(c *gin.Context) {
	var gpuStatus *service.GPUStatus
	var vllmMetrics *service.VLLMMetricsData
	if h.gpuMonitor != nil {
		gpuStatus = h.gpuMonitor.GetStatus()
		vllmMetrics = h.gpuMonitor.GetVLLMMetrics()
	}

	h.prometheus.UpdateGPUMetrics(prometheusGPUStatus(gpuStatus))
	h.prometheus.UpdateVLLMMetrics(prometheusVLLMMetrics(vllmMetrics))

	if h.metrics != nil {
		healthInfo := h.metrics.GetComprehensiveHealthScore(gpuStatus, vllmMetrics)
		if overall, ok := healthInfo["overall"].(float64); ok {
			h.prometheus.SetHealthScore(overall)
		}
	}
	if h.scheduler != nil {
		for _, model := range h.scheduler.GetAvailableModels() {
			h.prometheus.SetModelStatus(model, h.scheduler.GetModelService(model), h.scheduler.IsModelRunning(model))
			h.prometheus.SetActiveRequests(model, h.scheduler.GetActiveRequests(model))
		}
	}

	h.prometheus.Handler().ServeHTTP(c.Writer, c.Request)
}

func (h *HealthHandler) GetMetricsMetadata(c *gin.Context) {
	c.JSON(http.StatusOK, h.prometheus.GetMetricsDict())
}

func (h *HealthHandler) GetHealthHistory(c *gin.Context) {
	limit := 60
	if v := c.Query("count"); v != "" {
		if n, err := fmt.Sscanf(v, "%d", &limit); err == nil && n > 0 {
			_ = n
		}
	}

	gpuHistory := h.metrics.GetGPUHistory(limit)
	history := make([]gin.H, 0)

	for _, entry := range gpuHistory {
		memUtil := entry.MemoryUtilization
		temp := entry.Temperature
		score := 100.0 - float64(memUtil)*0.5 - float64(temp)*0.5
		if temp > 85 {
			score -= float64(temp-85) * 5
		}
		if memUtil > 90 {
			score -= float64(memUtil-90) * 10
		}
		score = max(0, min(100, score))

		alertCount := 0
		if temp > 85 || memUtil > 90 {
			alertCount = 1
		}
		if temp > 95 || memUtil > 95 {
			alertCount = 2
		}

		statusStr := "healthy"
		if score < 70 {
			statusStr = "degraded"
		}
		if score < 50 {
			statusStr = "unhealthy"
		}

		history = append(history, gin.H{
			"timestamp":    entry.Timestamp,
			"health_score": score,
			"status":       statusStr,
			"alert_count":  alertCount,
		})
	}

	if len(history) == 0 {
		gpuStatus := h.gpuMonitor.GetStatus()
		vllmMetrics := h.gpuMonitor.GetVLLMMetrics()
		healthInfo := h.metrics.GetComprehensiveHealthScore(gpuStatus, vllmMetrics)
		gpuAlerts := h.metrics.GetGPUAlerts(gpuStatus)
		alertCount := 0
		if gpuAlerts != nil {
			alertCount = len(gpuAlerts)
		}
		history = append(history, gin.H{
			"timestamp":    time.Now().Format(time.RFC3339),
			"health_score": healthInfo["overall"],
			"status":       healthInfo["status"],
			"alert_count":  alertCount,
		})
	}

	c.JSON(http.StatusOK, history)
}

File: app-controller/.__codex_upload__/health.py
from fastapi import APIRouter
from typing import Optional
from datetime import datetime

from core.deps import gpu_monitor, metrics, cache_service, prometheus, scheduler
from core.logger import setup_logger

health_router = APIRouter()
logger = setup_logger()


def _status_from_score(score: float) -> str:
    if score >= 90:
        return "healthy"
    if score >= 70:
        return "degraded"
    if score >= 50:
        return "warning"
    return "critical"


def _score_from_gpu_history(entry: dict) -> Optional[float]:
    score = entry.get("health_score")
    if isinstance(score, (int, float)):
        return round(float(score), 2)

    temp = entry.get("temperature")
    mem_util = entry.get("memory_utilization")
    if mem_util is None:
        total_mem = entry.get("total_memory") or 0
        used_mem = entry.get("used_memory") or 0
        if total_mem:
            mem_util = (used_mem / total_mem) * 100

    if not isinstance(temp, (int, float)) or not isinstance(mem_util, (int, float)):
        return None

    temp_score = max(0, 100 - max(0, temp - 85) * 5)
    mem_score = max(0, 100 - max(0, mem_util - 90) * 10)
    return round(max(0, min(100, temp_score * 0.5 + mem_score * 0.5)), 2)


def _build_fallback_health(reason: str):
    return {
        "status": "critical",
        "timestamp": datetime.now().isoformat(),
        "health_score": 0,
        "details": {
            "overall": 0,
            "status": "critical",
            "service": 0,
            "gpu_overall": 0,
            "vllm_inference": 0,
            "response_time": 0,
            "alerts": [f"health check fallback: {reason}"],
        },
    }


@health_router.get("/health")
async def health_check(refresh: Optional[bool] = False):
    cache_key = "api:health"

    if not refresh:
        cached = cache_service.get(cache_key)
        if cached is not None:
            return cached

    try:
        gpu_status = gpu_monitor.get_gpu_status()
        vllm_metrics = gpu_monitor.get_vllm_metrics()
        health_info = metrics.get_comprehensive_health_score(gpu_status, vllm_metrics)
        result = {
            "status": health_info["status"],
            "timestamp": datetime.now().isoformat(),
            "health_score": health_info["overall"],
            "details": health_info
        }
        cache_service.set(cache_key, result, ttl_seconds=5)
    except Exception as exc:
        logger.error(f"health_check failed: {exc}")
        result = _build_fallback_health(str(exc))
    return result


@health_router.get("/health/detailed")
async def health_check_detailed():
    cache_key = "api:health:detailed"
    cached = cache_service.get(cache_key)
    if cached is not None:
        return cached

    try:
        gpu_status = gpu_monitor.get_gpu_status()
        vllm_metrics = gpu_monitor.get_vllm_metrics()
        health_info = metrics.get_comprehensive_health_score(gpu_status, vllm_metrics)

        gpu_avail = gpu_status is not None and gpu_status.get("status") == "available"
        gpu_utilization = 0
        gpu_temp = 0
        gpu_mem_used_pct = 0
        if gpu_status:
            gpu_utilization = gpu_status.get("utilization", 0)
            gpu_temp = gpu_status.get("temperature", 0)
            gpu_mem_used_pct = gpu_status.get("memory_utilization", 0)

        vllm_running = vllm_metrics is not None and vllm_metrics.get("vllm_available", False)
        vllm_active_requests = vllm_metrics.get("running_requests", 0) if vllm_metrics else 0

        result = {
            "overall_score": health_info.get("overall", 0),
            "status": health_info.get("status", "unknown"),
            "checks": {
                "gpu": {
                    "available": gpu_avail,
                    "utilization": gpu_utilization,
                    "temperature": gpu_temp,
                    "memory_used_pct": gpu_mem_used_pct,
                },
                "go_backend": {
                    "reachable": True,
                    "response_time_ms": 0,
                },
                "python_backend": {
                    "reachable": True,
                    "response_time_ms": 0,
                },
                "vllm_service": {
                    "running": vllm_running,
                    "active_requests": vllm_active_requests,
                },
                "redis": {
                    "available": True,
                    "connected": True,
                },
            },
            "alert_reasons": health_info.get("alerts") or metrics.get_gpu_alerts(gpu_status) or [],
            "timestamp": datetime.now().isoformat(),
        }
        cache_service.set(cache_key, result, ttl_seconds=5)
    except Exception as exc:
        logger.error(f"health_check_detailed failed: {exc}")
        fallback = _build_fallback_health(str(exc))
        result = {
            "overall_score": 0,
            "status": fallback["status"],
            "checks": {
                "gpu": {"available": False, "utilization": 0, "temperature": 0, "memory_used_pct": 0},
                "go_backend": {"reachable": False, "response_time_ms": 0},
                "python_backend": {"reachable": False, "response_time_ms": 0},
                "vllm_service": {"running": False, "active_requests": 0},
                "redis": {"available": False, "connected": False},
            },
            "alert_reasons": [f"health check failed: {str(exc)}"],
            "timestamp": fallback["timestamp"],
        }
    return result


@health_router.get("/metrics")
async def get_prometheus_metrics():
    try:
        gpu_status = gpu_monitor.get_gpu_status()
        vllm_metrics = gpu_monitor.get_vllm_metrics()
        prometheus.update_gpu_metrics(gpu_status)

        health_info = metrics.get_comprehensive_health_score(gpu_status, vllm_metrics)
        prometheus.set_health_score(health_info["overall"])

        for model_name in scheduler.get_available_models():
            prometheus.set_model_status(
                model_name,
                scheduler.get_model_service(model_name) or "",
                scheduler.is_model_running(model_name)
            )
            prometheus.set_active_requests(model_name, scheduler.get_active_requests(model_name))
    except Exception as exc:
        logger.error(f"get_prometheus_metrics failed: {exc}")

    return prometheus.generate_metrics()


@health_router.get("/metrics/metadata")
async def get_metrics_metadata():
    return prometheus.get_metrics_dict()


@health_router.get("/health/history")
async def get_health_history(count: int = 60):
    history = []
    try:
        gpu_history = gpu_monitor.get_gpu_history(count)
        for entry in gpu_history:
            ts = entry.get("timestamp", "")
            score = _score_from_gpu_history(entry)
            if score is not None:
                history.append({
                    "timestamp": ts,
                    "health_score": score,
                    "status": _status_from_score(score),
                    "alert_count": len(entry.get("alert_reasons") or []),
                    "source": "gpu_history",
                })
    except Exception as exc:
        logger.error(f"get_health_history failed: {exc}")

    if not history:
        try:
            gpu_status = gpu_monitor.get_gpu_status()
            vllm_metrics = gpu_monitor.get_vllm_metrics()
            health_info = metrics.get_comprehensive_health_score(gpu_status, vllm_metrics)
            history.append({
                "timestamp": datetime.now().isoformat(),
                "health_score": health_info.get("overall", 0),
                "status": health_info.get("status", "unknown"),
                "alert_count": len(health_info.get("alerts") or []),
                "source": "current",
            })
        except Exception as exc:
            logger.error(f"get_health_history fallback failed: {exc}")

    return history
